File: plugins/plugin-linear/typescript/src/actions/deleteIssue.ts
import {
  type Action,
  type ActionResult,
  type HandlerCallback,
  type HandlerOptions,
  type IAgentRuntime,
  logger,
  type Memory,
  ModelType,
  type State,
} from "@elizaos/core";
import { deleteIssueTemplate } from "../generated/prompts/typescript/prompts.js";
import type { LinearService } from "../services/linear";
import type { DeleteIssueParameters } from "../types/index.js";

export const deleteIssueAction: Action = {
  name: "DELETE_LINEAR_ISSUE",
  description: "Delete (archive) an issue in Linear",
  similes: [
    "delete-linear-issue",
    "archive-linear-issue",
    "remove-linear-issue",
    "close-linear-issue",
  ],

  examples: [
    [
      {
        name: "User",
        content: {
          text: "Delete issue ENG-123",
        },
      },
      {
        name: "Assistant",
        content: {
          text: "I'll archive issue ENG-123 for you.",
          actions: ["DELETE_LINEAR_ISSUE"],
        },
      },
    ],
    [
      {
        name: "User",
        content: {
          text: "Remove COM2-7 from Linear",
        },
      },
      {
        name: "Assistant",
        content: {
          text: "I'll archive issue COM2-7 in Linear.",
          actions: ["DELETE_LINEAR_ISSUE"],
        },
      },
    ],
    [
      {
        name: "User",
        content: {
          text: "Archive the bug report BUG-456",
        },
      },
      {
        name: "Assistant",
        content: {
          text: "I'll archive issue BUG-456 for you.",
          actions: ["DELETE_LINEAR_ISSUE"],
        },
      },
    ],
  ],

  async validate(runtime: IAgentRuntime, _message: Memory, _state?: State): Promise<boolean> {
    const apiKey = runtime.getSetting("LINEAR_API_KEY");
    return !!apiKey;
  },

  async handler(
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
    _options?: HandlerOptions,
    callback?: HandlerCallback
  ): Promise<ActionResult> {
    try {
      const linearService = runtime.getService<LinearService>("linear");
      if (!linearService) {
        throw new Error("Linear service not available");
      }

      const content = message.content.text;
      if (!content) {
        const errorMessage = "Please specify which issue to delete.";
        await callback?.({
          text: errorMessage,
          source: message.content.source,
        });
        return {
          text: errorMessage,
          success: false,
        };
      }

      let issueId: string;

      const params = _options?.parameters as DeleteIssueParameters | undefined;
      if (params?.issueId) {
        issueId = params.issueId;
      } else {
        const prompt = deleteIssueTemplate.replace("{{userMessage}}", content);

        const response = await runtime.useModel(ModelType.TEXT_LARGE, {
          prompt: prompt,
        });

        if (!response) {
          throw new Error("Failed to extract issue identifier");
        }

        try {
          const cleanedResponse = response
            .replace(/^```(?:json)?\n?/, "")
            .replace(/\n?```$/, "")
            .trim();
          const parsed = JSON.parse(cleanedResponse);

          issueId = parsed.issueId;
          if (!issueId) {
            throw new Error("Issue ID not found in parsed response");
          }
        } catch (parseError) {
          logger.warn("Failed to parse LLM response, falling back to regex parsing:", parseError);

          const issueMatch = content.match(/(\w+-\d+)/);
          if (!issueMatch) {
            const errorMessage = "Please specify an issue ID (e.g., ENG-123) to delete.";
            await callback?.({
              text: errorMessage,
              source: message.content.source,
            });
            return {
              text: errorMessage,
              success: false,
            };
          }

          issueId = issueMatch[1];
        }
      }

      const issue = await linearService.getIssue(issueId);
      const issueTitle = issue.title;
      const issueIdentifier = issue.identifier;

      logger.info(`Archiving issue ${issueIdentifier}: ${issueTitle}`);

      await linearService.deleteIssue(issueId);

      const successMessage = `✅ Successfully archived issue ${issueIdentifier}: "${issueTitle}"\n\nThe issue has been moved to the archived state and will no longer appear in active views.`;
      await callback?.({
        text: successMessage,
        source: message.content.source,
      });

      return {
        text: `Archived issue ${issueIdentifier}: "${issueTitle}"`,
        success: true,
        data: {
          issueId: issue.id,
          identifier: issueIdentifier,
          title: issueTitle,
          archived: true,
        },
      };
    } catch (error) {
      logger.error("Failed to delete issue:", error);
      const errorMessage = `❌ Failed to delete issue: ${error instanceof Error ? error.message : "Unknown error"}`;
      await callback?.({
        text: errorMessage,
        source: message.content.source,
      });
      return {
        text: errorMessage,
        success: false,
      };
    }
  },
};

File: plugins/plugin-linear/python/src/elizaos_plugin_linear/actions/delete_issue.py
import json
import logging
import re
from typing import Any

from elizaos_plugin_linear.actions.base import (
    ActionExample,
    ActionResult,
    HandlerCallback,
    Memory,
    RuntimeProtocol,
    State,
    create_action,
)
from elizaos_plugin_linear.services.linear import LinearService

logger = logging.getLogger(__name__)

DELETE_ISSUE_TEMPLATE = """Given the user's request to delete/archive a Linear issue, extract the issue identifier.

User request: "{user_message}"

Extract and return ONLY a JSON object:
{{
  "issueId": "The issue identifier (e.g., ENG-123, COM2-7)"
}}

Return only the JSON object, no other text."""


async def validate(
    runtime: RuntimeProtocol,
    _message: Memory,
    _state: State | None = None,
) -> bool:
    try:
        api_key = runtime.get_setting("LINEAR_API_KEY")
        return bool(api_key)
    except Exception:
        return False


async def handler(
    runtime: RuntimeProtocol,
    message: Memory,
    _state: State | None = None,
    options: dict[str, Any] | None = None,
    callback: HandlerCallback | None = None,
) -> ActionResult:
    try:
        linear_service: LinearService = runtime.get_service("linear")
        if not linear_service:
            raise RuntimeError("Linear service not available")

        content = message.get("content", {}).get("text", "")
        if not content:
            error_message = "Please specify which issue to delete."
            if callback:
                await callback(
                    {"text": error_message, "source": message.get("content", {}).get("source")}
                )
            return {"text": error_message, "success": False}

        issue_id: str = ""

        if options and options.get("issueId"):
            issue_id = str(options["issueId"])
        else:
            prompt = DELETE_ISSUE_TEMPLATE.format(user_message=content)
            response = await runtime.use_model("TEXT_LARGE", {"prompt": prompt})

            if not response:
                raise RuntimeError("Failed to extract issue identifier")

            try:
                cleaned = re.sub(r"^```(?:json)?\n?", "", response)
                cleaned = re.sub(r"\n?```$", "", cleaned).strip()
                parsed = json.loads(cleaned)

                issue_id = parsed.get("issueId", "")
                if not issue_id:
                    raise ValueError("Issue ID not found")

            except json.JSONDecodeError:
                logger.warning("Failed to parse LLM response, falling back to regex")

                issue_match = re.search(r"(\w+-\d+)", content)
                if not issue_match:
                    error_message = "Please specify an issue ID (e.g., ENG-123) to delete."
                    if callback:
                        await callback(
                            {
                                "text": error_message,
                                "source": message.get("content", {}).get("source"),
                            }
                        )
                    return {"text": error_message, "success": False}

                issue_id = issue_match.group(1)

        issue = await linear_service.get_issue(issue_id)
        issue_title = issue.get("title", "Unknown")
        issue_identifier = issue.get("identifier", issue_id)

        logger.info(f"Archiving issue {issue_identifier}: {issue_title}")

        await linear_service.delete_issue(issue_id)

        success_message = f'✅ Successfully archived issue {issue_identifier}: "{issue_title}"\n\nThe issue has been moved to the archived state.'
        if callback:
            await callback(
                {"text": success_message, "source": message.get("content", {}).get("source")}
            )

        return {
            "text": f'Archived issue {issue_identifier}: "{issue_title}"',
            "success": True,
            "data": {
                "issueId": issue.get("id"),
                "identifier": issue_identifier,
                "title": issue_title,
                "archived": True,
            },
        }

    except Exception as error:
        logger.error(f"Failed to delete issue: {error}")
        error_message = f"❌ Failed to delete issue: {error}"
        if callback:
            await callback(
                {"text": error_message, "source": message.get("content", {}).get("source")}
            )
        return {"text": error_message, "success": False}


delete_issue_action = create_action(
    name="DELETE_LINEAR_ISSUE",
    description="Delete (archive) an issue in Linear",
    similes=["delete-linear-issue", "archive-linear-issue", "remove-linear-issue"],
    examples=[
        [
            ActionExample(name="User", content={"text": "Delete issue ENG-123"}),
            ActionExample(
                name="Assistant",
                content={
                    "text": "I'll archive issue ENG-123 for you.",
                    "actions": ["DELETE_LINEAR_ISSUE"],
                },
            ),
        ],
    ],
    validate=validate,
    handler=handler,
)
